### PR TITLE
[tune] OptunaSearch: check compatibility of search space with evaluated_rewards

### LINKED ISSUE #18587: [tune]: OptunaSearch define-by-run space incompatible with evaluated_rewards
### What is the problem?

I'm using Ray Tune for running hyperparameter optimization using OptunaSearch as a search algorithm. There are many options to configure the algorithm. Especially, there is a variety of different ways to describe the search space. When using the define-by-run way, it is not possible to provide evaluated_rewards.

```
Traceback (most recent call last):
  File "tune.py", line 20, in <module>
    algo = OptunaSearch(
  File ".../.venv/lib/python3.8/site-packages/ray/tune/suggest/optuna.py", line 243, in __init__
    self._setup_study(mode)
  File ".../.venv/lib/python3.8/site-packages/ray/tune/suggest/optuna.py", line 270, in _setup_study
    self.add_evaluated_point(point, reward)
  File ".../.venv/lib/python3.8/site-packages/ray/tune/suggest/optuna.py", line 419, in add_evaluated_point
    trial = ot.trial.create_trial(
  File ".../.venv/lib/python3.8/site-packages/optuna/_experimental.py", line 68, in new_func
    return func(*args, **kwargs)  # type: ignore
  File ".../.venv/lib/python3.8/site-packages/optuna/trial/_frozen.py", line 618, in create_trial
    trial._validate()
  File ".../.venv/lib/python3.8/site-packages/optuna/trial/_frozen.py", line 337, in _validate
    if set(self.params.keys()) != set(self.distributions.keys()):
AttributeError: 'function' object has no attribute 'keys'
```

The reason is that Optuna's interface expects a distribution (dict), when creating a trial. But the OptunaSearch class passes its own search space, which in this case is a callable. Optuna's validation fails. I guess right now the user must provide a distribution dict when using evaluated_rewards.

I would be happy to open a PR to:
- raise an exception explaining the incompatibilities before calling optuna
- add a hint for this issue in the docs
- add working examples to the docs

Ray 1.6.0
Optuna 2.9.1
Python 3.8.10

### Reproduction (REQUIRED)

```python
import optuna
from ray import tune
from ray.tune.suggest.optuna import OptunaSearch


def objective(x):
    return (x - 2)**2


def trainable(config):
    x = config["x"]
    loss = objective(x)
    tune.report(objective_value=loss)


points = [{"x": 4.0}, {"x": 1.0}]

rewards = [objective(config["x"]) for config in points]

algo = OptunaSearch(
    # WORKS
    # space={"x": optuna.distributions.UniformDistribution(-10, 10)},
    # FAILS
    space=lambda trial: {"x": trial.suggest_float("x", -10, 10)},
    metric="objective_value",
    mode="min",
    points_to_evaluate=points,
    evaluated_rewards=rewards,
)

algo = tune.suggest.ConcurrencyLimiter(algo, max_concurrent=1)

analysis = tune.run(
    trainable,
    metric="objective_value",
    mode="min",
    search_alg=algo,
    num_samples=10,
)

best = analysis.get_best_config("objective_value", "min")

print(best)
```

It is worth mentioning, that the above script works fine with define-by-run when no evaluated_rewards is passed to OptunaSearch.

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by carlassmann**
## Why are these changes needed?

OptunaSearch accepts both distribution dictionary or define-by-run as a definition for the search space. The latter should not be used, when `evaluated_rewards` is provided. Creation of trials in Optuna only allows the distribution strategy.

This PR adds docs for this case and provides a meaningful error when define-by-run is used with evaluated_rewards

## Related issue number
Closes #18587 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/

- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I need some guidance how to run the tests. Is there guidance on how to execute the tests? Installing the dependencies inside the python folder and running `python -m unittest` fails to execute the tests: `ImportError: Failed to import test module: ray`.

### CODE REVIEW

**Comment by ccssmnn** [comment]
@Yard1

**Comment by Yard1** [comment]
@ccssmnn Lint is still failing with `python/ray/tune/tests/test_searchers.py:310:80: E501 line too long (83 > 79 characters)`, can you check? Also, please merge the latest upstream master into your branch, there have been some CI issues that were fixed.

As for running tests locally, you should create a new virutalenv, follow [instructions on how to setup ray python dev environment](https://docs.ray.io/en/latest/development.html#building-ray-python-only) and then `python/requirements/requirements_default.txt` as well as `python/requirements/tune/requirements_tune.txt`. That should allow you to run tests without issues with `pytest TEST_FILE_NAME.py`

File: python/ray/tune/suggest/optuna.py [hunk 1]
```diff
@@ -54,7 +54,6 @@ class _OptunaTrialSuggestCaptor:
     `suggest_` callables with a function capturing the returned value,
     which will be saved in the ``captured_values`` dict.
     """
-
     def __init__(self, ot_trial: OptunaTrial) -> None:
```
Comment (Yard1) [inline python/ray/tune/suggest/optuna.py:57]: Can we leave an empty line here?

File: python/ray/tune/suggest/optuna.py [hunk 2]
```diff
@@ -126,6 +125,12 @@ class OptunaSearch(Searcher):
             needing to re-compute the trial. Must be the same length as
             points_to_evaluate.
 
+            ..warning::
+                When using evaluated_rewards, the search space `space` must
```
Comment (Yard1) [inline python/ray/tune/suggest/optuna.py:129]:
```suggestion
                When using ``evaluated_rewards``, the search space ``space`` must
```

File: python/ray/tune/suggest/optuna.py [hunk 3]
```diff
@@ -126,6 +125,12 @@ class OptunaSearch(Searcher):
...
+                be provided as a :class:`dict` with parameter names as keys and
+                ``optuna.distributions``. The define-by-run definition is not
+                yet supported.
```
Comment (Yard1) [inline python/ray/tune/suggest/optuna.py:132]:
```suggestion
                ``optuna.distributions`` instances as values. The define-by-run
                 search space definition is not yet supported with this functionality.
```

File: python/ray/tune/suggest/optuna.py [hunk 4]
```diff
@@ -178,14 +183,58 @@ def define_search_space(trial: optuna.Trial):
 
         tune.run(trainable, search_alg=optuna_search)
 
+    You can pass configs that will be evaluated first using
```
Comment (Yard1) [inline python/ray/tune/suggest/optuna.py:186]: Is the indentation here correct?

**Review by Yard1 — CHANGES_REQUESTED** [review]
There is a lot of unrelated code being reformatted. Can you ensure that you are using the right yapf version for `scripts/format.sh`, and that only related code is commited? Thanks!

File: python/ray/tune/suggest/optuna.py [hunk 5]
```diff
@@ -402,6 +452,11 @@ def add_evaluated_point(self,
                     cls=self.__class__.__name__,
                     metric=self._metric,
                     mode=self._mode))
+        if callable(self._space):
+            raise TypeError(
+                "Define-by-run function passed in `space` argument is not"
+                "yet supported when using `evaluated_rewards`. Please provide"
+                "a 'OptunaDistribution' dict.")
```
Comment (Yard1) [inline python/ray/tune/suggest/optuna.py:459]:
```suggestion
                "Define-by-run function passed in `space` argument is not "
                "yet supported when using `evaluated_rewards`. Please provide "
                "an 'OptunaDistribution' dict.")
```

**Review by Yard1 — APPROVED** [review]
Looks good, thanks! Just one last nit. Feel free to mark it as ready for review.

**Comment by ccssmnn** [comment]
Thanks @Yard1 !

I'm still running into errors executing the test locally though.

My steps:

1. create a GitHub CodeSpace from this PR
2. remove .venv which is created by vscode
3. create a new venv with `python -m venv venv`
4. activate venv `source venv/bin/activate`
5. install the latest ray wheels with pip command as suggested in the URL you provided
6. run `python python/ray/setup-dev.py`
7. install dependencies with `pip install -r python/requirements.txt`
8. run test `pytest python/ray/tune/tests/test_searchers.py`

Results in the following error:

```
(venv) codespace ➜ /workspaces/ray (pr/ccssmnn/18625 ✗) $ pytest python/ray/tune/tests/test_searchers.py 
Test session starts (platform: linux, Python 3.8.6, pytest 5.4.3, pytest-sugar 0.9.4)
rootdir: /workspaces/ray/python
plugins: lazy-fixture-0.6.3, sugar-0.9.4, rerunfailures-10.1, anyio-3.3.1, asyncio-0.15.1, timeout-1.4.2
collecting ... 
―――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― ERROR collecting ray/tune/tests/test_searchers.py ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――
python/ray/tune/tests/test_searchers.py:5: in <module>
    import ray
venv/lib/python3.8/site-packages/ray/__init__.py:89: in <module>
    from ray.worker import (  # noqa: E402,F401
venv/lib/python3.8/site-packages/ray/worker.py:24: in <module>
    import ray._private.parameter
venv/lib/python3.8/site-packages/ray/_private/parameter.py:11: in <module>
    class RayParams:
venv/lib/python3.8/site-packages/ray/_private/parameter.py:153: in RayParams
    worker_setup_hook=ray_constants.DEFAULT_WORKER_SETUP_HOOK,
E   AttributeError: module 'ray.ray_constants' has no attribute 'DEFAULT_WORKER_SETUP_HOOK'

=================================================================================== short test summary info ===================================================================================
FAILED python/ray/tune/tests/test_searchers.py - AttributeError: module 'ray.ray_constants' has no attribute 'DEFAULT_WORKER_SETUP_HOOK'
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

Results (0.42s):
```

This may be irrelevant for this PR, but it can be a blocker for future PRs.

**Comment by Yard1** [comment]
@ccssmnn Hey, sorry to keep you waiting. The workflow you have posted is pretty much exactly what I am doing. Not sure why it would cause such an error for you. Can you try again?

I will merge master into your branch to make CI pass.

File: python/ray/tune/suggest/optuna.py [hunk 6]
```diff
@@ -367,6 +417,11 @@ def add_evaluated_point(self,
                     cls=self.__class__.__name__,
                     metric=self._metric,
                     mode=self._mode))
+        if callable(self._space):
+            raise TypeError(
+                "Define-by-run function passed in `space` argument is not "
+                "yet supported when using `evaluated_rewards`. Please provide "
+                "an 'OptunaDistribution' dict.")
```
Comment (krfricke) [inline python/ray/tune/suggest/optuna.py:424]:
```suggestion
                "an `OptunaDistribution` dict or pass a Ray Tune "
                "search space to `tune.run()`.")
```

**Review by krfricke — APPROVED** [review]
Thanks!

**Pull request merged**